public/app/features/plugins/plugin_loader.ts

```diff
@@ -187,52 +187,55 @@ export function importDataSourcePlugin(meta: DataSourcePluginMeta): Promise<Gene
   });
 }
 
-// Only successfully loaded plugins are cached
-const importedAppPlugins: Record<string, AppPlugin> = {};
+// Cache for import promises to prevent duplicate imports
+const importPromises: Record<string, Promise<AppPlugin>> = {};
 
 export async function importAppPlugin(meta: PluginMeta): Promise<AppPlugin> {
   const pluginId = meta.id;
 
-  if (importedAppPlugins[pluginId]) {
-    return importedAppPlugins[pluginId];
+  // If import is already in progress or completed, return the existing promise
+  if (importPromises[pluginId] !== undefined) {
+    return importPromises[pluginId];
   }
 
-  throwIfAngular(meta);
+  importPromises[pluginId] = (async () => {
+    throwIfAngular(meta);
 
-  const pluginExports = await importPluginModule({
-    path: meta.module,
-    version: meta.info?.version,
-    pluginId: meta.id,
-    loadingStrategy: meta.loadingStrategy ?? PluginLoadingStrategy.fetch,
-    moduleHash: meta.moduleHash,
-    translations: meta.translations,
-  });
+    const pluginExports = await importPluginModule({
+      path: meta.module,
+      version: meta.info?.version,
+      pluginId: meta.id,
+      loadingStrategy: meta.loadingStrategy ?? PluginLoadingStrategy.fetch,
+      moduleHash: meta.moduleHash,
+      translations: meta.translations,
+    });
 
-  const { plugin = new AppPlugin() } = pluginExports;
-  plugin.init(meta);
-  plugin.meta = meta;
-  plugin.setComponentsFromLegacyExports(pluginExports);
+    const { plugin = new AppPlugin() } = pluginExports;
+    plugin.init(meta);
+    plugin.meta = meta;
+    plugin.setComponentsFromLegacyExports(pluginExports);
 
-  exposedComponentsRegistry.register({
-    pluginId,
-    configs: plugin.exposedComponentConfigs || [],
-  });
-  addedComponentsRegistry.register({
-    pluginId,
-    configs: plugin.addedComponentConfigs || [],
-  });
-  addedLinksRegistry.register({
-    pluginId,
-    configs: plugin.addedLinkConfigs || [],
-  });
-  addedFunctionsRegistry.register({
-    pluginId,
-    configs: plugin.addedFunctionConfigs || [],
-  });
+    exposedComponentsRegistry.register({
+      pluginId,
+      configs: plugin.exposedComponentConfigs || [],
+    });
+    addedComponentsRegistry.register({
+      pluginId,
+      configs: plugin.addedComponentConfigs || [],
+    });
+    addedLinksRegistry.register({
+      pluginId,
+      configs: plugin.addedLinkConfigs || [],
+    });
+    addedFunctionsRegistry.register({
+      pluginId,
+      configs: plugin.addedFunctionConfigs || [],
+    });
 
-  importedAppPlugins[pluginId] = plugin;
+    return plugin;
+  })();
```

Comment:
[nitpick]: I understand this is a temporary fix until we start work on https://github.com/grafana/grafana/issues/97106 but maybe we should break out this part into a separate async function and call that instead of using a self executing function?

Reply:
Yeah, that's a good point @hugohaggmark, will extract it out 👍

Reply:
Fixed in https://github.com/grafana/grafana/pull/107579/commits/063e31ac2a103686626b6130bb8df36960d32c0c.